gmso/core/views.py

```diff
@@ -0,0 +1,180 @@
+    PotentialFilters.UNIQUE_NAME_CLASS: get_name_or_class,
+    PotentialFilters.UNIQUE_EXPRESSION: lambda p: str(p.expression),
+    PotentialFilters.UNIQUE_PARAMETERS: get_parameters,
+}
+
+
+class TopologyPotentialView:
+    """A potential view based on different filters for a topology's potentials.
+
+    Parameters
+    ----------
+    iterator: typing.Iterator, required=True
+        An iterator of either topology sites or connections from which to extract potentials from
+
+    Parameters
+    ----------
+    filter_by: str or function, default=None
+        If provided, filter the collected potentials by some unique identifier
+        of a potential.
+
+    Notes
+    -----
+    If `filter_by` is provided and is a custom function, the collected potentials are
+    filtered on the basis of the return value of the `filter_by` function. A single potential
+    is passed to the filter_by function and it should return an identifier (should be hashable)
+    for that potential thus describing its uniqueness in the context of which the filter is being
+    used. Some simple examples are given below.
+
+    Examples
+    --------
+    To use a TopologyPotentialView, a Topology must have few sites with AtomTypes or BondTypes.
+
+    >>> from gmso.core.topology import Topology
+    >>> from gmso.core.atom import Atom
+    >>> from gmso.core.atom_type import AtomType
+    >>> top = Topology(name="ViewTopology")
+    >>> sites = [Atom(name=f"Atom_{j}") for j in range(10)]
+    >>> atom_type1 = AtomType(name='atom_type1')
+    >>> atom_type2 = AtomType(name='atom_type2')
+    >>> for site in sites:
+    ...     site.atom_type = atom_type1 if int(site.name[-1]) % 2 == 0 else atom_type2
+    ...     top.add_site(site)
+    >>> top.update_topology()
+    >>> for atom_type in top.atom_types:
+    ...     print(atom_type.name)
+    atom_type1
+    atom_type2
+    >>> top.get_index(atom_type2)
+    1
+
+    Notes
+    -----
+    The implementation of this class is inspired from networkx.classes.reportviews.NodeView by extending
+    the idea of a view with filteration capabilities. See the source for NodeView for further details
+
+    https://github.com/networkx/networkx/blob/12c1a00cd116701a763f7c57c230b8739d2ed085/networkx/classes/reportviews.py#L115-L279
+    """
+
+    def __init__(self, iterator, filter_by=None):
+        self.iterator = iterator
+        self.filter_by = filter_by
+
+    def __iter__(self):
+        yield from self.yield_view()
+
+    def index(self, item):
+        for j, potential in enumerate(self.yield_view()):
+            if potential is item:
+                return j
+
+    def _collect_potentials(self):
+        """Collect potentials from the iterator"""
+        visited = set()
+        for item in self.iterator:
+            potential = getattr(
+                item, potential_attribute_map[type(item)]
+            )  # Since this use is internal, KeyErrors N/A
+            if potential and id(potential) not in visited:
```

Comment:
For the above behavior, I think you would just get rid of this `if id(potential) not in visited` clause.